### ISSUE TITLE
Making instrumentation no-op on demand to reduce tracing overhead

### ISSUE THREAD

**Issue by kravnaut** [DataDog/dd-trace-py#5556]
### Summary of problem

Note: This is something we have separately engaged on with the support team as one of your customers but wanted to flag this here as well. 

I understand the motivation to always create spans and write them to the agent even if sampler said not to (or even when explicitly said not to via tag). I suppose you are doing this to support some metrics at agent level and force traces with errors to be collected (tail-based processing) even when sampling flag was set not to. Problem is this overhead is very significant and leaves folks like us with high fan-out/load with a choice of either completely disabling tracing or enabling tracing only on some of our fleet. While we were hoping to just trace a percentage of requests to get a reasonable benefit w/o all of the overhead.

What we have done so far is basically patch this Pin [method](https://github.com/DataDog/dd-trace-py/blob/b4449028209854654291cb762d02baf4bd1cf8cd/ddtrace/pin.py#L150) to return False if a contextvar flag has been configured. This allows us to manually check the sampler decision and raise this flag to block span creation. 

It would be ideal if this was first-class feature in ddtrace to mitigate the performance overhead. We expected that if sampling check did not pass then spans would be no-op.

Basically what we need is either a ddtrace config that honors unsampled traces to be truly no-op or provides a runtime flag we can put on and off to suppress tracer from being enabled in a specific context (i.e. a request for us)

### Which version of dd-trace-py are you using?

`ddtrace==1.4.4`

**Comment by brettlangdon** [comment]
@kravnaut thanks for reaching out.

We have a feature in beta at the moment which _might_ help your scenario.

Setting the `DD_TRACE_STATS_COMPUTATION_ENABLED=true` environment variable will instruct the ddtrace library to compute the stats for all traces and then only encode and flush kept traces to the trace agent.

This is not the same as what you are asking for since you are asking for unsampled spans to become a no-op. This will still compute and store all spans, but we will actually drop them in the tracing client.

We've found from testing this isn't a performance win for all scenarios, since the computation of the stats adds it's own overhead. So, I am a bit reluctant to say that this will resolve your problems, but it is definitely worth trying. If you have a very low sample rate, as long as the stats aggregation/encoding/flushing is lower overhead than encoding/flushing the spans then it should be a win.




There are two main challenges with no-op spans:

1. Today, stats are computed in the trace agent, so in order to get accurate metrics the spans (sampled or not) must make it to the trace agent. This is where the beta for computing stats in the tracing client comes in.
2. The sampling decision can change at any point during the lifetime of a span. Some products might override the sampling decision if they notice an error/anomaly that we think a user wants, or a user can manually override/set the sampling decision at any point.

If we can get 1 in a steady state/enabled by default, then I think the discussion around no-op spans becomes more interesting/realistic (personal opinion)

**Comment by kravnaut** [comment]
Our issue is that we have endpoints with high QPS but more importantly with substantial fan-out through usage of things like GraphQL, so the performance overhead really adds up both in CPU and latency. So while writing to the agent also contributes to network footprint, it's really the span creation logic that seems to contribute the most.

We can take a look at the beta feature but intuitively it doesn't sound like it will help our case I'm describing. Under our load patterns we would have been just fine with metrics of only sampled traces given that they would be fairly statistically significant. I would suggest that you consider offering such capability with the disclaimer that it may skew computed stats if the the sampled traces/spans are not representative of general load patterns.

**Comment by brettlangdon** [comment]
> So while writing to the agent also contributes to network footprint, it's really the span creation logic that seems to contribute the most.

👍🏻 makes sense. thanks for sharing.

**Comment by emmettbutler** [comment]
Manually doing what the Github Action was supposed to do (fixed in https://github.com/DataDog/dd-trace-py/pull/7835):

This issue has been automatically closed after six months of inactivity. If it's a feature request, it has been added to the maintainers' internal backlog and will be included in an upcoming round of feature prioritization. Please comment or reopen if you think this issue was closed in error.
